include/revng/BasicAnalyses/AdvancedValueInfo.h

```diff
@@ -561,60 +593,104 @@ class Expression {
-            return I;
-      return nullptr;
-    };
+  llvm::Instruction *isInteresting(const Operation &O) {
+    if (O.Range.isFullSet())
+      if (auto *I = dyn_cast<llvm::Instruction>(O.V))
+        if (isa<llvm::IntegerType>(I->getType()))
+          return I;
+    return nullptr;
+  };
+
+  /// For each instruction in the chain record the number of possible values
+  /// according to LVI.
+  void recordAVIValue(llvm::ArrayRef<llvm::Instruction *> Targets,
+                      llvm::LazyValueInfo &LVI,
+                      const llvm::DominatorTree &DT,
+                      PhiEdges &Edges,
+                      llvm::BasicBlock *StopAt,
+                      const detail::Graph &Graph) {
+    using namespace llvm;
+
+    if (Targets.size() == 0)
+      return;
+
+    BasicBlock *StartBB = Targets.back()->getParent();
+
+    const CFGEdge &FirstEdge = Edges.front();
+    revng_assert(FirstEdge.End == nullptr);
+    BasicBlock *EndBB = FirstEdge.Start;
+
+    BasicBlock *LimitedStartBB = StartBB;
+    if (not DT.dominates(StopAt, LimitedStartBB))
+      LimitedStartBB = StopAt;
+
+    SmallPtrSet<BasicBlock *, 4> IgnoreList = { StopAt };
+
+    auto Reachable = nodesBetweenReverse(EndBB, LimitedStartBB, &IgnoreList);
+    for (Instruction *I : Targets)
+      Reachable.insert(I->getParent());
+    Reachable.insert(EndBB);
+
+    // Note: ordering in reverse post order is more costly than beneficial
+    Reachable.erase(StartBB);
+    SmallVector<BasicBlock *, 8> ReachableVector{ StartBB };
+    for (BasicBlock *BB : Reachable)
+      ReachableVector.push_back(BB);
+
+    DisjointRanges::Analysis DR(ReachableVector,
+                                LVI,
+                                DT,
+                                Targets,
+                                Edges,
+                                Graph);
+    DR.initialize();
+    DR.run();
+
+    for (Operation &O : OperationsStack) {
+      if (auto *I = isInteresting(O)) {
+        O.Range = DR.get(I);
+        AVILogger << I << ": ";
+        O.Range.dump(AVILogger);
+        AVILogger << DoLog;
+        O.RangeSize = O.Range.size().getLimitedValue();
+      }
+    }
+  }
+
+  /// Use LVI to build an expression about \p V
+  ///
+  /// 1. Build a chain of single non-const-operand instructions until you find a
+  ///    phi or a load from a global variable.
+  /// 2. For each instruction in the chain record the number of possible values
+  ///    according to LVI.
+  /// 3. Iterate over the chain looking for the instruction associated with the
+  ///    smallest range.
+  llvm::Instruction *buildExpression(llvm::LazyValueInfo &LVI,
+                                     const llvm::DominatorTree &DT,
+                                     PhiEdges &Edges,
+                                     llvm::Value *V,
+                                     llvm::BasicBlock *StopAt,
+                                     const detail::Graph &Graph) {
```

Comment:
1. Do not introduce new terminology unless necessary. The concept of "chains" is unclear and overlapping with "expression". Rename `buildChains` to `buildExpression` and what currently is `buildExpression` to `analyzeValue` or `buildAndAnalyzeExpression`. Fix the docs too.
2. `recordAVIValue` does not record values, it's misleading. Rename to `computeOperationRanges`.

Reply:
ok i think i misunderstood parts of the code i was outlining, i renamed them as indicated